File: src/services/userService.js
const { v4: uuidv4 } = require('uuid');
const crypto = require('crypto');

/**
 * Hash a password using SHA-256
 * @param {string} password
 * @returns {string} hashed password
 */
const hashPassword = (password) => {
  return crypto.createHash('sha256').update(password).digest('hex');
};

// Seed users with default password 'password123'
const DEFAULT_PASSWORD_HASH = hashPassword('password123');

// In-memory data store
let users = [
  { id: '1', name: 'Alice Johnson', email: 'alice@example.com', role: 'admin', passwordHash: DEFAULT_PASSWORD_HASH, createdAt: '2024-01-01T00:00:00.000Z' },
  { id: '2', name: 'Bob Smith', email: 'bob@example.com', role: 'user', passwordHash: DEFAULT_PASSWORD_HASH, createdAt: '2024-01-02T00:00:00.000Z' },
  { id: '3', name: 'Carol White', email: 'carol@example.com', role: 'user', passwordHash: DEFAULT_PASSWORD_HASH, createdAt: '2024-01-03T00:00:00.000Z' },
];

/**
 * Get all users
 * @returns {Array} list of users
 */
const getAllUsers = () => {
  // Exclude passwordHash from returned objects for security
  return users.map(({ passwordHash: _, ...user }) => user);
};

/**
 * Get user by ID
 * @param {string} id
 * @returns {Object|null} user or null
 */
const getUserById = (id) => {
  const user = users.find((u) => u.id === id);
  if (!user) return null;
  const { passwordHash: _, ...userWithoutPassword } = user;
  return userWithoutPassword;
};

/**
 * Get user by email (internal method, includes passwordHash)
 * @param {string} email
 * @returns {Object|null} user or null
 */
const getUserByEmailInternal = (email) => {
  return users.find((u) => u.email === email) || null;
};

/**
 * Get user by email
 * @param {string} email
 * @returns {Object|null} user or null
 */
const getUserByEmail = (email) => {
  const user = getUserByEmailInternal(email);
  if (!user) return null;
  const { passwordHash: _, ...userWithoutPassword } = user;
  return userWithoutPassword;
};

/**
 * Create a new user
 * @param {Object} data - { name, email, role, password }
 * @returns {Object} created user
 * @throws {Error} if email already exists or required fields missing
 */
const createUser = (data) => {
  const name = data.name ? data.name.trim() : data.name;
  const email = data.email ? data.email.trim().toLowerCase() : data.email;
  const role = data.role || 'user';
  const password = data.password || 'password123'; // Default fallback password

  if (!name || !email) {
    throw new Error('Name and email are required');
  }

  if (getUserByEmailInternal(email)) {
    throw new Error('Email already exists');
  }

  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  if (!emailRegex.test(email)) {
    throw new Error('Invalid email format');
  }

  const validRoles = ['admin', 'user'];
  if (!validRoles.includes(role)) {
    throw new Error(`Role must be one of: ${validRoles.join(', ')}`);
  }

  const newUser = {
    id: uuidv4(),
    name,
    email,
    role,
    passwordHash: hashPassword(password),
    createdAt: new Date().toISOString(),
  };

  users.push(newUser);
  const { passwordHash: _, ...userWithoutPassword } = newUser;
  return userWithoutPassword;
};

/**
 * Update a user
 * @param {string} id
 * @param {Object} updates
 * @returns {Object|null} updated user or null
 */
const updateUser = (id, updates) => {
  const index = users.findIndex((u) => u.id === id);
  if (index === -1) return null;

  const { name, email, role, password } = updates;

  if (email && email !== users[index].email) {
    if (getUserByEmailInternal(email)) {
      throw new Error('Email already exists');
    }
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    if (!emailRegex.test(email)) {
      throw new Error('Invalid email format');
    }
  }

  if (role) {
    const validRoles = ['admin', 'user'];
    if (!validRoles.includes(role)) {
      throw new Error(`Role must be one of: ${validRoles.join(', ')}`);
    }
  }

  users[index] = {
    ...users[index],
    ...(name && { name: name.trim() }),
    ...(email && { email: email.trim().toLowerCase() }),
    ...(role && { role }),
    ...(password && { passwordHash: hashPassword(password) }),
    updatedAt: new Date().toISOString(),
  };

  const { passwordHash: _, ...userWithoutPassword } = users[index];
  return userWithoutPassword;
};

/**
 * Delete a user
 * @param {string} id
 * @returns {boolean} true if deleted, false if not found
 */
const deleteUser = (id) => {
  const index = users.findIndex((u) => u.id === id);
  if (index === -1) return false;
  users.splice(index, 1);
  return true;
};

/**
 * Verify a user's password
 * @param {string} email
 * @param {string} password
 * @returns {Object|null} User details without password if verified, else null
 */
const verifyPassword = (email, password) => {
  if (!email || !password) return null;
  const user = getUserByEmailInternal(email.trim().toLowerCase());
  if (!user) return null;

  const inputHash = hashPassword(password);
  if (user.passwordHash === inputHash) {
    const { passwordHash: _, ...userWithoutPassword } = user;
    return userWithoutPassword;
  }
  return null;
};

/**
 * Reset the data store (used in tests)
 */
const resetUsers = () => {
  users = [
    { id: '1', name: 'Alice Johnson', email: 'alice@example.com', role: 'admin', passwordHash: DEFAULT_PASSWORD_HASH, createdAt: '2024-01-01T00:00:00.000Z' },
    { id: '2', name: 'Bob Smith', email: 'bob@example.com', role: 'user', passwordHash: DEFAULT_PASSWORD_HASH, createdAt: '2024-01-02T00:00:00.000Z' },
    { id: '3', name: 'Carol White', email: 'carol@example.com', role: 'user', passwordHash: DEFAULT_PASSWORD_HASH, createdAt: '2024-01-03T00:00:00.000Z' },
  ];
};

module.exports = {
  getAllUsers,
  getUserById,
  getUserByEmail,
  createUser,
  updateUser,
  deleteUser,
  verifyPassword,
  resetUsers,
};

File: src/main/java/com/example/automation/service/UserService.java
package com.example.automation.service;

import com.example.automation.model.User;
import org.springframework.stereotype.Service;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.Instant;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

@Service
public class UserService {

    private final List<User> users = new CopyOnWriteArrayList<>();
    private final String DEFAULT_PASSWORD_HASH = hashPassword("password123");

    public UserService() {
        resetUsers();
    }

    private String hashPassword(String password) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] encodedhash = digest.digest(password.getBytes(StandardCharsets.UTF_8));
            StringBuilder hexString = new StringBuilder(2 * encodedhash.length);
            for (byte b : encodedhash) {
                String hex = Integer.toHexString(0xff & b);
                if (hex.length() == 1) {
                    hexString.append('0');
                }
                hexString.append(hex);
            }
            return hexString.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-256 algorithm not found", e);
        }
    }

    public List<User> getAllUsers() {
        return users.stream().map(this::removePassword).collect(Collectors.toList());
    }

    public User getUserById(String id) {
        return users.stream()
                .filter(u -> u.getId().equals(id))
                .findFirst()
                .map(this::removePassword)
                .orElse(null);
    }

    private User getUserByEmailInternal(String email) {
        return users.stream()
                .filter(u -> u.getEmail().equals(email))
                .findFirst()
                .orElse(null);
    }

    public User getUserByEmail(String email) {
        User user = getUserByEmailInternal(email);
        if (user == null)
            return null;
        return removePassword(user);
    }

    public User createUser(User data) {
        String name = data.getName() != null ? data.getName().trim() : null;
        String email = data.getEmail() != null ? data.getEmail().trim().toLowerCase() : null;
        String role = data.getRole() != null ? data.getRole() : "user";
        String password = data.getPasswordHash() != null ? data.getPasswordHash() : "password123";

        if (name == null || name.isEmpty() || email == null || email.isEmpty()) {
            throw new IllegalArgumentException("Name and email are required");
        }

        if (getUserByEmailInternal(email) != null) {
            throw new IllegalArgumentException("Email already exists");
        }

        if (!Pattern.matches("^[\\w-\\.]+@([\\w-]+\\.)+[\\w-]{2,4}$", email)) {
            throw new IllegalArgumentException("Invalid email format");
        }

        if (!role.equals("admin") && !role.equals("user")) {
            throw new IllegalArgumentException("Role must be one of: admin, user");
        }

        User newUser = new User();
        newUser.setId(UUID.randomUUID().toString());
        newUser.setName(name);
        newUser.setEmail(email);
        newUser.setRole(role);
        newUser.setPasswordHash(hashPassword(password));
        newUser.setCreatedAt(Instant.now().toString());

        users.add(newUser);
        return removePassword(newUser);
    }

    public User updateUser(String id, User updates) {
        Optional<User> optionalUser = users.stream().filter(u -> u.getId().equals(id)).findFirst();
        if (optionalUser.isEmpty())
            return null;

        User existingUser = optionalUser.get();

        String email = updates.getEmail();
        if (email != null && !email.equals(existingUser.getEmail())) {
            if (getUserByEmailInternal(email) != null) {
                throw new IllegalArgumentException("Email already exists");
            }
            if (!Pattern.matches("^[\\w-\\.]+@([\\w-]+\\.)+[\\w-]{2,4}$", email)) {
                throw new IllegalArgumentException("Invalid email format");
            }
            existingUser.setEmail(email.trim().toLowerCase());
        }

        String name = updates.getName();
        if (name != null) {
            existingUser.setName(name.trim());
        }

        String role = updates.getRole();
        if (role != null) {
            if (!role.equals("admin") && !role.equals("user")) {
                throw new IllegalArgumentException("Role must be one of: admin, user");
            }
            existingUser.setRole(role);
        }

        String password = updates.getPasswordHash(); // Assuming input passed as raw password in this field
        if (password != null) {
            existingUser.setPasswordHash(hashPassword(password));
        }

        existingUser.setUpdatedAt(Instant.now().toString());
        return removePassword(existingUser);
    }

    public boolean deleteUser(String id) {
        return users.removeIf(u -> u.getId().equals(id));
    }

    public User verifyPassword(String email, String password) {
        if (email == null || password == null)
            return null;
        User user = getUserByEmailInternal(email.trim().toLowerCase());
        if (user == null)
            return null;

        String inputHash = hashPassword(password);
        if (user.getPasswordHash().equals(inputHash)) {
            return removePassword(user);
        }
        return null;
    }

    public void resetUsers() {
        users.clear();
        User u1 = new User();
        u1.setId("1");
        u1.setName("Alice Johnson");
        u1.setEmail("alice@example.com");
        u1.setRole("admin");
        u1.setPasswordHash(DEFAULT_PASSWORD_HASH);
        u1.setCreatedAt("2024-01-01T00:00:00.000Z");
        User u2 = new User();
        u2.setId("2");
        u2.setName("Bob Smith");
        u2.setEmail("bob@example.com");
        u2.setRole("user");
        u2.setPasswordHash(DEFAULT_PASSWORD_HASH);
        u2.setCreatedAt("2024-01-02T00:00:00.000Z");
        User u3 = new User();
        u3.setId("3");
        u3.setName("Carol White");
        u3.setEmail("carol@example.com");
        u3.setRole("user");
        u3.setPasswordHash(DEFAULT_PASSWORD_HASH);
        u3.setCreatedAt("2024-01-03T00:00:00.000Z");
        users.add(u1);
        users.add(u2);
        users.add(u3);
    }

    private User removePassword(User user) {
        User copy = new User();
        copy.setId(user.getId());
        copy.setName(user.getName());
        copy.setEmail(user.getEmail());
        copy.setRole(user.getRole());
        copy.setCreatedAt(user.getCreatedAt());
        copy.setUpdatedAt(user.getUpdatedAt());
        return copy;
    }
}
